### DISCUSSION TITLE
MDF-e (Manifesto)

### DISCUSSION THREAD

**Discussion by leogregianin** [TadaSoftware/PyNFe#112] [Q&A] [answered]
Abri essa discussão sobre a branch [new_mdfe](https://github.com/TadaSoftware/PyNFe/tree/new_mdfe) para inclusão do MDF-e no projeto.

Exemplos de código de como utilizá-la estão no [wiki](https://github.com/TadaSoftware/PyNFe/wiki#mdf-e) e no [PR original](https://github.com/TadaSoftware/PyNFe/pull/79).

A classe [Comunicacao](https://github.com/TadaSoftware/PyNFe/blob/01f53eaee07353b8adf67489e47558ef1e856e19/pynfe/processamento/comunicacao.py#L20) da branch new_mdfe ficou abstraída e criadas novas classes para cada fim em ```nfe.py```, ```mdfe.py``` e ```nfse.py```, com isso, deve haver adaptações nos ```imports``` dos métodos para quem já utiliza o PyNFe, exemplo:

De:
```python
from pynfe.processamento.comunicacao import ComunicacaoSefaz
from pynfe.processamento.comunicacao import ComunicacaoNfse
```

Para:
```python
from pynfe.processamento.nfe import ComunicacaoNFe
from pynfe.processamento.nfse import ComunicacaoNfse
```

O que acham?

**Comment by juniortada** [comment, marked as answer]
Bom dia @leogregianin!

Eu vi a modificação, ficou muito bom.

Aquela modificação que você tinha sugerido de separar os impostos em aquivos,
também é muito boa. Eu estou com pouco tempo para me dedicar ao PyNFe,
mas assim que tiver tempo ou quando for necessário uma manutenção maior nesta parte do código
vou implementar.

**Reply by leogregianin** [reply]
Bom dia, legal @juniortada.

Vou trabalhar para melhorar a separação dos grupos de impostos da NFe/NFCe.

E sobre a branch do Manifesto, quais os planos? Após o merge dela na master podemos incluir novos serviços como CT-e, etc.

**Reply by snpxbr** [reply]
Boa tarde @leogregianin 

Existe alguma previsão de quando será incluído o módulo de CTe?

**Reply by leogregianin** [reply]
Boa tarde @snpxbr, realmente não tem previsão. Precisamos fazer testes unitários para o que já existe e mergear a branch [new_mdfe](https://github.com/TadaSoftware/PyNFe/tree/new_mdfe) para master.

**Comment by leogregianin** [comment]
Atualizando esse tópico: O MDF-e já está na branch principal e os exemplos de uso estão no [wiki](https://github.com/TadaSoftware/PyNFe/wiki#mdf-e).

**Reply by inacioloy** [reply]
Caro @leogregianin testei com os exemplos da wiki, porém ao testar em em homologação dá falha no schema do XML, quando verifico o XML gerado em https://dfe-portal.svrs.rs.gov.br/Mdfe/ValidadorXML, realmente ocorre erro na validação de schema. Vocês estão conseguindo validar o XML da MDF-e /  emitir MDF-e normalmente? Segue anexo o xml gerado no meu teste
[mdfe_homolog.zip](https://github.com/TadaSoftware/PyNFe/files/9411873/mdfe_homolog.zip)

**Reply by leogregianin** [reply]
Olá @inacioloy, esse XML que anexou é o envelope SOAP para envio para o Webservice, dessa forma ele não vai validar no Validador do Portal do MDFe, você pode extrair somente o XML que será validado sim.

Por [aqui](http://lucrorural.com.br) já emitimos bastante MDF-es com integração com a lib PyNFe.

Temos testes automáticos do XML gerado, o código-fonte está [aqui](https://github.com/TadaSoftware/PyNFe/blob/main/tests/test_mdfe_serializacao.py), e é validado com os _schemas_ XSD.

**Reply by inacioloy** [reply]
Obrigado pela resposta @leogregianin, realmente me equivoquei ao verificar o XML com o envelope SOAP, ontem ainda consegui validar o schema, muito grato. Outra dúvida, vocês utilizam qual lib para gerar a versão impressa (Documento Auxiliar do Manifesto)?

**Reply by leogregianin** [reply]
@inacioloy para relatórios, impressão de NF-e e MDF-e temos uma solução própria baseada nos templates do Django e algumas outras libs como [django-weasyprint](https://github.com/fdemmer/django-weasyprint), [qrcode](https://github.com/lincolnloop/python-qrcode) e [python-barcode](https://github.com/WhyNotHugo/python-barcode).
